crates/runtime/src/traphandlers.rs

```diff
@@ -603,13 +735,90 @@ mod tls {
     }
 }
 
+/// A module for handling faults on Fuchsia
+#[cfg(target_os = "fuchsia")]
+mod fuchsia {
+    use super::*;
+
+    // Be very careful here. There is a contract with the stack in setup_fault_handler
+    pub extern "C" fn fault_handler(
+        regs: *const sys::zx_thread_state_general_regs_t,
+        report: *const sys::zx_exception_report_t,
+    ) {
+        let handled = tls::with(|info| {
+            let info = match info {
+                Some(info) => info,
+                None => return false,
+            };
+
+            let jmp_buf = unsafe {
+                info.handle_trap((*regs).rip as *const _, false, |handler| handler(*report))
+            };
+
+            if jmp_buf.is_null() {
+                return false;
+            } else if jmp_buf as usize == 1 {
+                return true;
+            } else {
+                unsafe { Unwind(jmp_buf) }
+            }
+        });
+    }
+
+    fn align_16(x: usize) -> usize {
+        ((x + 16 - 1) / 16) * 16
+    }
+
+    /// Intended to be called on the exception handling thread. Sets up the stack and updates the
+    /// rip for the faulted thread to call the fault_handler function when it is resumed.
+    pub fn setup_fault_handler(
+        regs: &mut sys::zx_thread_state_general_regs_t,
+        report: &sys::zx_exception_report_t,
+    ) {
+        let fault_addr = regs.rip;
+        regs.rsp -= regs.rsp % 16;
+
+        // put the register info on the stack
+        regs.rsp -= align_16(std::mem::size_of::<sys::zx_thread_state_general_regs_t>()) as u64;
+        let info_regs: sys::zx_thread_state_general_regs_t = regs.clone();
+        unsafe {
+            (*(regs.rsp as *mut _)) = info_regs;
+        }
+        regs.rdi = regs.rsp;
+
+        // exception report info
+        regs.rsp -= align_16(std::mem::size_of::<sys::zx_exception_report_t>()) as u64;
+        let report: sys::zx_exception_report_t = report.clone();
+        unsafe {
+            (*(regs.rsp as *mut _)) = report;
+        }
+        regs.rsi = regs.rsp;
+
+        // reasonable place to ret
+        regs.rsp -= 8;
+        unsafe { (*(regs.rsp as *mut _)) = fault_addr + 1 };
```

Comment:
I think there may need to be some special tidbits when resuming the original thread here? I think you probably want to resume the instruction itself rather than one byte past, and additionally I think something here needs to clean up the stack allocation for the context info, but I'm not sure that anything currently does?

Reply:
Thinking some more about this, is there a syscall or a standard method for "resume this thread given this thread context?" Because you basically want to always resume the exact same instruction in the exact same context (all registers preserved which currently aren't I think), but sometimes you'll unregister the trap handling logic.